### ISSUE TITLE
Can't get max-batch-size to work

### ISSUE THREAD

**Issue by whillas** [triton-inference-server/server#2192]
**Description**

Can not get the model to load with a max_batch_size specified in the `config.pbtxt` file

My model config input (as auto generated by Triton):
```json
  "input": [
    {
      "name": "image_tensor",
      "data_type": "TYPE_FP32",
      "format": "FORMAT_NONE",
      "dims": [
        -1,
        456,
        456,
        3
      ],
      "is_shape_tensor": false,
      "allow_ragged_batch": false
    }
  ],
```
If i set the `max_batch_size` in the config.pbtxt as the [documentation states](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/protobuf_api/model_config.proto.html#_CPPv4N9inference11ModelConfig14max_batch_sizeE):

> A max_batch_size value > 0 indicates that batching is allowed and so the model expects the input tensors to have an additional initial dimension for the batching that is not specified in the input (for example, if the model supports batched inputs of 2-dimensional tensors then the model configuration will specify the input shape as [ X, Y ] but the model will expect the actual input tensors to have shape [ N, X, Y ])

```
name: "fixed-batch"
platform: "tensorflow_savedmodel"
max_batch_size: 8
input {
    name: "image_tensor"
    data_type: TYPE_FP32
    dims: 456
    dims: 456
    dims: 3
}
```
I get the following error:

> Internal: unable to autofill for 'fixed-batch', configuration specified max-batch 8 but model signature does not support batching

I get this if I
- Add the `dim: -1` back into the config.pbtxt
- export the model with a fix batch size of 8 i.e. `dims: 8`

**Triton Information**

Official 20.9 image

**To Reproduce**

Tensorflow Efficientnet-b5 model with batch dimension 

**Expected behavior**

To be able to limit the max batch size and for the model to load.

**Comment by deadeyegoodwin** [comment]
If you used triton to auto-complete a configuration then it will set max_batch_size == 1 if batching is supported for the model (and you can then copy that config and increase max_batch_size if you want). What is the entire generated configuration?

Or, with the adjustments you describe it could be that you just need to not use --strict-model-config=false after you have modified the configuration, because you don't want triton to try to auto-complete your modified configuration.

**Comment by whillas** [comment]
We are using `--strict-model-config=false` atm, that's how i got the model input config above. It seems no matter what combination of `config.pbtxt` options i use I can't get Triton to handle the batching. Here is the full config:
```
{
  "name": "efficientnet-b5",
  "platform": "tensorflow_savedmodel",
  "backend": "tensorflow",
  "version_policy": {
    "latest": {
      "num_versions": 1
    }
  },
  "max_batch_size": 0,
  "input": [
    {
      "name": "image_tensor",
      "data_type": "TYPE_FP32",
      "format": "FORMAT_NONE",
      "dims": [
        -1,
        456,
        456,
        3
      ],
      "is_shape_tensor": false,
      "allow_ragged_batch": false
    }
  ],
  "output": [
    {
      "name": "logits",
      "data_type": "TYPE_FP32",
      "dims": [
        -1,
        52
      ],
      "label_filename": "",
      "is_shape_tensor": false
    },
    {
      "name": "classes",
      "data_type": "TYPE_STRING",
      "dims": [
        52
      ],
      "label_filename": "",
      "is_shape_tensor": false
    }
  ],
  "batch_input": [],
  "batch_output": [],
  "optimization": {
    "priority": "PRIORITY_DEFAULT",
    "input_pinned_memory": {
      "enable": true
    },
    "output_pinned_memory": {
      "enable": true
    }
  },
  "instance_group": [
    {
      "name": "groups_1_1_1_1_1_1_1",
      "kind": "KIND_GPU",
      "count": 1,
      "gpus": [
        0
      ],
      "profile": []
    }
  ],
  "default_model_filename": "model.savedmodel",
  "cc_model_filenames": {},
  "metric_tags": {},
  "parameters": {},
  "model_warmup": []
}
```
What should the `config.pbtxt` look like to have a batch size of 8 on this model?

**Comment by deadeyegoodwin** [comment]
You can see that triton sets max-batch-size == 0 which means that it doesn't know how to batch for this model. It looks like the reason for that is that the "classes" output does not have an initial batch dimension.

**Comment by whillas** [comment]
yeah, its a variable batch size.
OK, so whats the solution? Do i re-export the model with a fixed batch size? In that case what should the `config.pbtxt` look like?
OR should I export the model with no batch dimension and again what should the `config.pbtxt` look like.

I want to avoid the dynamic batching in the latter case as, for Tensorflow models, it seems to consume all the memory à la #2109 which is _why_ i'm trying to fix the batch size in the first place (to stop this memory leak in TF).

**Comment by deadeyegoodwin** [comment]
As far as I can tell you model as shown in the above config has a batch dimension on all inputs and outputs except 'classes'. That causes Triton to say "I can't tell if this model supports batching in the first dimension, so I will turn off my dynamic batching support" (by setting max_batch_size == 0). Do you want Triton to be able to dynamically batch for your model? If so you need to change your model to have a batch dimension for 'classes'. If you don't want dynamic batching you can leave the model as is, or even change the model so that the -1 first dimension is some constant value (but then every request would need to be exactly that size).

**Comment by whillas** [comment]
so when should `max_batch_size > 0`?

**Comment by deadeyegoodwin** [comment]
https://github.com/triton-inference-server/server/blob/master/docs/model_configuration.md#inputs-and-outputs

**Comment by deadeyegoodwin** [comment]
Closing, reopen if additional information.
